module #3. REST API Advanced/src/main/java/com/epam/esm/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,117 @@
+package com.epam.esm.controller;
+
+import com.epam.esm.assembler.GiftCertificateAssembler;
+import com.epam.esm.controller.constraint.FilterConstraint;
+import com.epam.esm.dto.GiftCertificateDto;
+import com.epam.esm.dto.group.OnPersist;
+import com.epam.esm.service.GiftCertificateService;
+import org.springframework.data.domain.Page;
+import org.springframework.hateoas.Link;
+import org.springframework.hateoas.PagedModel;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.MultiValueMap;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+import javax.validation.constraints.Min;
+
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
+
+/**
+ * Controller class to manage {@link GiftCertificateDto Gift Certificates}.
+ */
+@RestController
+@RequestMapping("api/certificates")
+@Validated
+public class GiftCertificateController {
+
+    private final GiftCertificateService giftCertificateService;
+
+    private final GiftCertificateAssembler giftCertificateAssembler;
+
+    public GiftCertificateController(GiftCertificateService giftCertificateService, GiftCertificateAssembler giftCertificateAssembler) {
+        this.giftCertificateService = giftCertificateService;
+        this.giftCertificateAssembler = giftCertificateAssembler;
+    }
+
+    /**
+     * Method to get all {@link GiftCertificateDto Gift Certificates} with Pagination and Filtering options.
+     *
+     * @param page number of the page
+     * @param size number of items in a page
+     * @param filterParams filter parameters to apply
+     * @return a {@link PagedModel} which contains all the {@link GiftCertificateDto Gift Certificates} with given filter options
+     */
+    @GetMapping()
```

Comment:
First of all "()" is redundant since you dont pass any arguments,
Secondly, its better to specify accepts/produces params at @GetMapping, @PostMapping annotations etc.
So will be something like that @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)